examples/energy-management-app/energy-management-common/src/EnergyEvseTargetsStore.cpp

```diff
@@ -0,0 +1,415 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <EnergyEvseTargetsStore.h>
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <lib/support/SafeInt.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::EnergyEvse;
+
+// using namespace chip::app::Clusters::EnergyEvse::Attributes;
+
+EvseTargetIteratorImpl * EvseTargetsDelegate::GetEvseTargetsIterator()
+{
+    if (mEvseTargetsIterator == nullptr)
+    {
+        // TODO replace this without using new
+        EvseTargetsDelegate & dg = *this;
+        mEvseTargetsIterator     = new EvseTargetIteratorImpl(dg);
+    }
+    return mEvseTargetsIterator;
+};
+EvseTargetsDelegate::~EvseTargetsDelegate()
+{
+    if (mEvseTargetsIterator)
+    {
+        delete mEvseTargetsIterator;
+    }
+}
+
+size_t EvseTargetIteratorImpl::Count()
+{
+    return static_cast<size_t>(mTargetEntryVector.size());
+}
+bool EvseTargetIteratorImpl::Next(EvseTargetEntry & entry)
+{
+    if (mTargetEntryIndex < mTargetEntryVector.size())
+    {
+        entry = mTargetEntryVector[mTargetEntryIndex];
+        mTargetEntryIndex++;
+        return true;
+    }
+    return false;
+}
+
+void EvseTargetIteratorImpl::Release()
+{
+    mTargetEntryIndex = 0;
+    mTargetEntryVector.clear();
+}
+
+CHIP_ERROR EvseTargetIteratorImpl::Load()
+{
+    size_t targetsSize = EvseTargetsDelegate::MaxTargetEntrySize();
+    ReturnErrorOnFailure(mDelegate.Load(mTargetEntryVector, targetsSize));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EvseTargetsDelegate::Init(PersistentStorageDelegate * targetStore)
+{
+    ChipLogProgress(AppServer, "EVSE: Initializing EvseTargetsDelegate");
+    VerifyOrReturnError(targetStore != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    mpTargetStore = targetStore;
+
+    return CHIP_NO_ERROR;
+}
+CHIP_ERROR EvseTargetsDelegate::LoadCounter(size_t & count, size_t & targetsSize)
+{
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    size_t len = MaxTargetEntryCounterSize();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(!backingBuffer.Calloc(len), CHIP_ERROR_NO_MEMORY);
+    uint16_t length = static_cast<uint16_t>(len);
+
+    CHIP_ERROR err =
+        mpTargetStore->SyncGetKeyValue(DefaultStorageKeyAllocator::EvseTargetEntryCounter().KeyName(), backingBuffer.Get(), length);
+    if (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND)
+    {
+        return CHIP_NO_ERROR;
```

Comment:
But the out params did not get set, did they?  How is this supposed to work?

Reply:
I have set the values of we return here.